### PR TITLE
Hub client

### PR DESCRIPTION
**Pull request by crazy-max**
Basic Hub API client to authenticate through docker CLI. Inspired by `docker/hub-tool` hub client but auth config is retrieved from the daemon.

Maybe we could mutualize the [current hub client in hub-tool](https://github.com/docker/hub-tool/tree/main/internal/hub) and move the package `internal/hub` to `pkg/hub`? (@silvin-lubecki)

cc @eunomie

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

File: cli/hub/client.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		creds := struct {
+			Token string `json:"token"`
+		}{}
+		if err = json.Unmarshal(buf, &creds); err != nil {
+			return err
+		}
+		c.username = authconfig.Username
+		c.token = creds.Token
+		return nil
+	} else if resp.StatusCode == http.StatusUnauthorized {
```
Comment (eunomie) [inline cli/hub/client.go:82]: nit: you don't need the `else`

File: cli/hub/client.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		}
+
+		// Check if 2FA is enabled and needs a second authentication
+		if response2FA.Detail != secondFactorDetailMessage {
+			return fmt.Errorf(response2FA.Detail)
+		}
+
+		// TODO: Retrieve two factor token
+		c.username = authconfig.Username
+		//c.token =
```
Comment (eunomie) [inline cli/hub/client.go:98]: Maybe I'm wrong but I think we don't want to deal with 2FA here. And we shouldn't.
If user has 2FA, the accessible credential should be a PAT.
If user don't have 2FA, no prob.

**Review by eunomie — COMMENTED** [review]
Can be interesting to have a `IsLoggedIn` method. So we can know the current state.
If I'm right, if there's no auth config for the URL, then it's not authenticated

File: cli/hub/client.go (hunk 2 above)
Comment (crazy-max) [reply cli/hub/client.go:98]: @eunomie SGTM

**Comment by crazy-max** [comment]
> Can be interesting to have a `IsLoggedIn` method. So we can know the current state.
> If I'm right, if there's no auth config for the URL, then it's not authenticated

Yeah will add this method

File: cli/hub/repositories.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	"time"
+)
+
+const (
+	// repositoriesURL path to the Hub API listing the repositories
+	repositoriesURL = "/v2/repositories/%s/"
+)
+
+// Repository represents a Docker Hub repository
+type Repository struct {
```
Comment (eunomie) [inline cli/hub/repositories.go:17]: If it's just to test, maybe we don't need that and we can only check if the response from Hub is OK (no error and json). We don't need to validate the response is a real repositories one.

**Pull request closed without merging**